src/assets/CollectibleDetectionController.ts

```diff
@@ -145,12 +145,20 @@ export class CollectibleDetectionController extends BaseController<
       /* istanbul ignore if */
       do {
         const api = this.getOwnerCollectiblesApi(address, offset);
-        response = await timeoutFetch(
-          api,
-          openSeaApiKey ? { headers: { 'X-API-KEY': openSeaApiKey } } : {},
-          15000,
-        );
-        const collectiblesArray = await response.json();
+        try {
+          response = await timeoutFetch(api, {}, 15000);
+        } catch {
```

Comment:
By omitting the `error` being caught here, we're suppressing all errors thrown in this `try` block. If something unexpected gets thrown, something indicative of a dev problem on our end (like invalid params, broken auth, malformed URL, etc.) we'd never know.

Reply:
We are [currently swallowing errors in this same method](https://github.com/MetaMask/controllers/blob/main/src/assets/CollectibleDetectionController.ts#L159) but I agree we shouldn't be doing this.

Reply:
ping here @Gudahtt

Reply:
Generally we should let "programmer errors"/"unexpected errors" bubble up to be caught by Sentry if possible.

For errors that we expect to happen in some circumstances, like those caused by a misbehaving or offline server or a network issue, there is no one-size-fits-all approach. I'm not sure what would be most appropriate here. I didn't have anything specific in mind. It sounds like you've settled on logging for at least some of them.

Reply:
I don't currently know of any errors we can deem "expected" or more of warnings that we can safely ignore. So I suppose we should log all errors until we become aware of some that might fit that description. In `controllers` how do we send errors to sentry?  The `safelyExecute` utility method that we wrap a bunch of stuff in seems to just `console.error` all errors. Do these get sent to sentry somehow?

Reply:
This is a fetch call, so we should expect it to fail for any network error or server error or timeout, for starters.

Reply:
There is no way to directly send errors to Sentry from this package aside from bubbling them up to the project. The `safetyExecute` utility should be considered deprecated; it swallows errors completely by default, and if enabled it logs them but only to the console, so we don't see them in prod.

Reply:
> This is a fetch call, so we should expect it to fail for any network error or server error or timeout, for starters.

Sorry I don't think I follow what you think we should do on these types of errors. Are you saying if we get certain status codes we should swallow them?

Reply:
And is [this a good example](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/metametrics.js#L86) of the current pattern/method we should be using to bubble errors from a controller to sentry (minus the exceptionsToFilter)? @Gudahtt

Reply:
No, I didn't express any opinion on what to do about those errors. Those were examples of "expected errors", you said you couldn't think of any examples.

We need to ensure unexpected errors are logged to Sentry, so that we can catch our own bugs. Errors that we expect to happen, operation errors like network failures, should be treated on a case-by-case basis in whatever way we think is appropriate for the situation. As I said, I'm not sure what would be most appropriate here.

Reply:
>And is [this a good example](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/controllers/metametrics.js#L86) of the current pattern/method we should be using to bubble errors from a controller to sentry (minus the exceptionsToFilter)

By "bubble up" what I mean is throw the exception, letting it proceed up the call stack until caught by a `catch` block or by Sentry's global unhandled error handler. That's generally what I'd recommend for any errors we don't anticipate.

Reply:
See this guide for more error handling advice: https://www.joyent.com/node-js/production/design/errors

Reply:
On the options for handling fetch errors, here are a few basic options:

* Swallow the error completely
  * Pros: no runtime cost
  * Cons: Hides potential problems
  If we _really, really_ don't care about the error, this might be appropriate. e.g. If we're using errors as control flow, like for validation, we don't need to log to console if `validateJson(badJson)` throws. But I can't think of an example where this makes sense for a fetch call. We always want a fetch call to succeed.
* Print the error to console
  * Pros: Highlights problem in development
  * Cons: Problem remains hidden in prod, and could flood console if it's too high-frequency.
  This is a good solution in cases where we _really_ don't care about the error, we just want to move on. The log is there for visibility. We can set it as debug level if it's a very mundane scenario that we usually don't care about, but usually warn or error level is more appropriate for a fetch error.
* Log the failure with a metric event
  * Pros: Lets us track frequency of failures in prod, increasing our confidence in the API/service or alerting us to prod failures, or failures that affect certain user segments
  * Cons: More complex to setup, and risks flooding Segment with events if frequency is too high.

Additional options that could be combined with the above strategies:
* We could try retrying the error (with or without logging it), maybe with an exponential backoff strategy to give the servers time to recover if they're down due to high traffic or a DDoS attack. We'd want to be careful that the error actually is something that might succeed upon retry though, or risk wasting time, bandwidth, and increasing server costs.
* We could limit error logging (to console or Segment) by time frame, such that we only log once per minute/hour/day or something like that, as a safeguard against flooding the console or Segment with events.
* We could use some sort of error sampling or aggregation technique. e.g. report 1st, 10th, 100th, and 1000th error, or report the number of errors per day, or report once per day if there has been "downtime" for a particular API.
* We could use browser APIs to detect internet connectivity, so that we don't even try if the user is offline. This would ensure any error logs/events we get are for more interesting problems. We don't care when the user is out of WiFi range, but we might care if the service tends to be blocked by corporate proxies or doesn't work with certain ISPs, etc.

Reply:
I could see value in doing this to reduce the ever growing noise in our logs:
> We could use browser APIs to detect internet connectivity, so that we don't even try if the user is offline. This would ensure any error logs/events we get are for more interesting problems.

In terms of retrying, this method is already called within a polling cycle and the implications for failure are relatively low impact, we simply won't detect new NFTs that OpenSea has indexed to this account until the service is back up when we try again.

I guess my vote is to log fetch errors to the console since we will likely be alerted to these issues through codefi's monitoring of this proxy server which they are managing. Alternatively we could log them to segment so we could get some metrics about how frequently this impacting users?

Reply:
Does this make sense to you @Gudahtt?

Reply:
Yep! Either of those plans make sense to me

Reply:
@Gudahtt here's an attempt to address some of your concerns: https://github.com/MetaMask/controllers/pull/805/commits/b21e4e5a7138e44b50cd79c37e1f7374831e6464

Regarding 
> We could use browser APIs to detect internet connectivity, so that we don't even try if the user is offline. 

I see [navigator.Online](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/onLine), what do think of pulling this API in as part of more widely used utility method to check before making any network requests?

Reply:
ping here when you get a chance @Gudahtt

Reply:
Totally know you're doing much more important stuff but going to keep pinging you just in case haha @Gudahtt

Reply:
Another friendly ping here @Gudahtt

Reply:
>I see [navigator.Online](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/onLine), what do think of pulling this API in as part of more widely used utility method to check before making any network requests?

Is that API supported by React Native, or a polyfill that we're already using? I wasn't sure how feasible using that is at this layer.

Reply:
> Is that API supported by React Native, or a polyfill that we're already using?

Good call out. 
This API is available for the extension at this layer but is almost certainly not supported in React Native... Not seeing a polyfill or any alternative? Any ideas on this @gantunesr or @sethkfman?